builder/virtualbox/common/step_ssh_key_pair.go

```diff
@@ -0,0 +1,118 @@
+package common
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"runtime"
+
+	"github.com/hashicorp/packer/common/uuid"
+	"github.com/hashicorp/packer/helper/communicator"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/helper/ssh"
+	"github.com/hashicorp/packer/packer"
+)
+
+// StepSshKeyPair executes the business logic for setting the SSH key pair in
+// the specified communicator.Config.
+type StepSshKeyPair struct {
+	Debug        bool
+	DebugKeyPath string
+	Comm         *communicator.Config
+}
+
+func (s *StepSshKeyPair) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	if s.Comm.SSHPassword != "" {
+		return multistep.ActionContinue
+	}
+
+	ui := state.Get("ui").(packer.Ui)
+
+	if s.Comm.SSHPrivateKeyFile != "" {
+		ui.Say("Using existing SSH private key for the communicator...")
+		privateKeyBytes, err := s.Comm.ReadSSHPrivateKeyFile()
+		if err != nil {
+			state.Put("error", err)
+			return multistep.ActionHalt
+		}
+
+		kp, err := ssh.NewKeyPairBuilder().
+			SetPrivateKey(privateKeyBytes).
+			SetName(fmt.Sprintf("packer_%s", uuid.TimeOrderedUUID())).
+			Build()
+		if err != nil {
+			state.Put("error", err)
+			return multistep.ActionHalt
+		}
+
+		s.Comm.SSHPrivateKey = privateKeyBytes
+		s.Comm.SSHKeyPairName = kp.Name()
+		s.Comm.SSHTemporaryKeyPairName = kp.Name()
+		s.Comm.SSHPublicKey = kp.PublicKeyAuthorizedKeysLine(ssh.UnixNewLine)
+
+		return multistep.ActionContinue
+	}
+
+	if s.Comm.SSHAgentAuth {
+		ui.Say("Using local SSH Agent to authenticate connections for the communicator...")
+		return multistep.ActionContinue
+	}
+
+	ui.Say("Creating ephemeral key pair for SSH communicator...")
+
+	kp, err := ssh.NewKeyPairBuilder().
+		SetName(fmt.Sprintf("packer_%s", uuid.TimeOrderedUUID())).
+		Build()
+	if err != nil {
+		state.Put("error", fmt.Errorf("Error creating temporary keypair: %s", err))
+		return multistep.ActionHalt
+	}
+
+	s.Comm.SSHKeyPairName = kp.Name()
+	s.Comm.SSHTemporaryKeyPairName = kp.Name()
+	s.Comm.SSHPrivateKey = kp.PrivateKeyPemBlock()
+	s.Comm.SSHPublicKey = kp.PublicKeyAuthorizedKeysLine(ssh.UnixNewLine)
+	s.Comm.SSHClearAuthorizedKeys = true
+
+	ui.Say(fmt.Sprintf("Created ephemeral SSH key pair of type %s", kp.Description()))
+
+	// If we're in debug mode, output the private key to the working
+	// directory.
+	// TODO: It would be better if the file was 'chmod' before writing
+	//  the key to the disk - or if umask was set before creating the file.
+	if s.Debug {
+		ui.Message(fmt.Sprintf("Saving communicator private key for debug purposes: %s", s.DebugKeyPath))
+		f, err := os.Create(s.DebugKeyPath)
+		if err != nil {
+			state.Put("error", fmt.Errorf("Error saving debug key: %s", err))
+			return multistep.ActionHalt
+		}
+		defer f.Close()
+
```

Comment:
I wrote the TODO because I copy-pasta'd from the `googlecompute` builder (I think?). I can fix it here, and add the TODO other places... There is no consistent way of handling this at the moment throughout the code base.

Reply:
Yep, that's a straightforward fix. I'm guessing that @stephen-fox left it that way (with TODO) to provoke this conversation. If fixing the file permission race condition in just one place is the agreed-upon path, we can do that.

Reply:
The call has been updated to use `os.OpenFile()`. The subsequent `os.File.Chmod()` has been removed.